src/win32-syscall.c

```diff
@@ -0,0 +1,1651 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \file
+ *
+ * \author Jacob Masen-Smith <jacob@evengx.com>
+ *
+ * Isolation for WMI/COM functionality
+ *
+ * References:
+ * https://msdn.microsoft.com/en-us/library/aa390421(v=vs.85).aspx
+ * https://blogs.msdn.microsoft.com/ndis/2015/03/21/mapping-from-ndis-oids-to-wmi-classes/
+ * https://stackoverflow.com/questions/1431103/how-to-obtain-data-from-wmi-using-a-c-application
+ * https://docs.microsoft.com/en-us/windows-hardware/drivers/network/oid-tcp-offload-parameters
+ * https://wutils.com/wmi/root/wmi/ms_409/msndis_tcpoffloadcurrentconfig/
+ * https://docs.microsoft.com/en-us/windows-hardware/drivers/network/oid-tcp-offload-current-config
+ * https://wutils.com/wmi/root/wmi/msndis_tcpoffloadparameters/
+ */
+
+#ifdef OS_WIN32
+
+#include <inttypes.h>
+#include <stdbool.h>
+
+// clang-format off
+#include <winsock2.h>
+#include <windows.h>
+#include <wbemidl.h>
+#include <strsafe.h>
+#include <ntddndis.h>
+#include <ws2ipdef.h>
+#include <iphlpapi.h>
+// clang-format on
+
+#include "util-debug.h"
+#include "util-device.h"
+#include "util-unittest.h"
+
+#include "win32-syscall.h"
+
+/**
+ * \brief return only the GUID portion of the name
+ */
+static const char *StripPcapPrefix(const char *pcap_dev)
+{
+    return strchr(pcap_dev, '{');
+}
+
+/**
+ * \brief get the adapter address list, which includes IP status/details
+ *
+ * Clients MUST FREE the returned list to avoid memory leaks.
+ */
+uint32_t Win32GetAdaptersAddresses(IP_ADAPTER_ADDRESSES **pif_info_list)
+{
+    DWORD err = NO_ERROR;
+    IP_ADAPTER_ADDRESSES *if_info_list;
+
+    ULONG size = 0;
+    err = GetAdaptersAddresses(AF_UNSPEC, 0, NULL, NULL, &size);
+    if (err != ERROR_BUFFER_OVERFLOW) {
+        return err;
+    }
+    if_info_list = malloc((size_t)size);
+    if (if_info_list == NULL) {
+        return ERROR_NOT_ENOUGH_MEMORY;
+    }
+    err = GetAdaptersAddresses(AF_UNSPEC, 0, NULL, if_info_list, &size);
+    if (err != NO_ERROR) {
+        free(if_info_list);
+        return err;
+    }
+
+    *pif_info_list = if_info_list;
+    return NO_ERROR;
+}
+
+uint32_t Win32FindAdapterAddresses(IP_ADAPTER_ADDRESSES *if_info_list,
+                                   const char *adapter_name,
+                                   IP_ADAPTER_ADDRESSES **pif_info)
+{
+    DWORD ret = NO_ERROR;
+    adapter_name = StripPcapPrefix(adapter_name);
+    *pif_info = NULL;
+
+    for (IP_ADAPTER_ADDRESSES *current = if_info_list; current != NULL;
+         current = current->Next) {
+
+        /* if we find the adapter, return that data */
+        if (strncmp(adapter_name, current->AdapterName, strlen(adapter_name)) ==
+            0) {
+
+            *pif_info = current;
+            break;
+        }
+    }
+
+    if (*pif_info == NULL) {
+        ret = ERROR_NOT_FOUND;
+    }
+
+    return ret;
+}
+
+#if NTDDI_VERSION < NTDDI_VISTA

```

Comment:
does this still make sense? Is there something older than Vista we'd want to support? I'm thinking there is not. As this is new functionality we can easily set a minimum supported version to be something recent. (e.g. just Win10 + latest Server?)

Reply:
That was a matter of assumption I made based on [this discussion](https://github.com/OISF/suricata/pull/3339#discussion_r181546625) that XP compatibility was still desirable.

Reply:
To be specific, the minimum Windows OS this functionality requires is Vista, for both the WMI APIs and WinDivert (which relies on Windows Filtering Platform introduced in Vista.)